waiter/docs/service-references.md

```diff
@@ -0,0 +1,46 @@
+# Service References
+
+To process a http request, Waiter needs to resolve a request to a service description.
+This service description tells Waiter how to locate the service instance to use to process the current request.
+
+Waiter uses request headers to construct the service description using the `ServiceDescriptionBuilder`.
+It is possible for the service to be reachable from various combinations of headers:
+  1. token - single or multiple
+  1. on-the-fly services without tokens
+  1. on-the-fly services that extend tokens, e.g. run-as-requester
+
+## Service GC
+
+By default, Waiter relies on the service idle timeout to GC services after periods of inactivity (not receiving requests).
+However, services reachable only via references, e.g. tokens, can be GC-ed eagerly if the reference has been updated.
+
+When a service description is constructed from a request, the service references are also updated.
+These references are available as the `:reference-type->entry` key in the descriptor.
+The `reference-type->entry` is a map where the keys represent a `:type` parameter, e.g.
+a service created with `x-waiter-token: foo` header on a request will have the value:
+```
+  {:token {:sources [{"token" "foo" "version" "v1"}]}}
```

Comment:
Why are we using strings for the "`token"` and `"version"` keys?

Reply:
How are multiple tokens handled?  Is that modeled differently from various tokens resolving to the same service?

Reply:
How does this map change if the same service was to be resolved directly without a token?